### DISCUSSION TITLE
How can you expand context that is passed in via an output

### DISCUSSION THREAD

**Discussion by Ajibola** [community/community#25355] [Actions]
<p>When passing data via the env as a string, the context is expanded, but if the env variable is populated via a file output, the context is not expanded.</p>
<p>this example below works fine</p>
<pre><code class="lang-auto">- name: Send Message
      env:
        MSG_TEMPLATE: '
        {
          "text": "*${{ github.workflow }}*\n`${{ job.status }}`"
        }
     run: echo ${{env.MSG_TEMPLATE}}
</code></pre>
<p>but if changed to the one below, it doesn’t work</p>
<pre><code class="lang-auto">id: gettemplate
run: echo ::set-output name=msg::$(cat template.json)
....
run: echo ${{ steps.gettemplate.outputs.msg }}
</code></pre>
<p>the first scenario described above outputs</p>
<blockquote>
<p>{  text : “<em>Workflow Name</em>\n<code>succeeds</code>” }</p>
</blockquote>
<p>the second scenario outputs</p>
<blockquote>
<p>{ “text”: “*<span class="math">{{ github.workflow }}*\n`</span>{{ job.status }}`” }</p>
</blockquote>
<p>How can I achieve the same results by passing the string as a “parameter”?</p>

**Comment by Yanjingzhu** [comment]
<p>According to the logs , when you set the json content with context directly as the output variable value, it could be evaluated.<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/1/127b60e15cfa9c5dda853f318b39722cf8b05652.png" data-download-href="/uploads/short-url/2DuVKTps9wZUcr2BkVbaqq92HJ0.png?dl=1" title="evaluate" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181090584-97b6e82a-3b0c-48bd-be85-d153aa4fde63.png" alt="evaluate" data-base62-sha1="2DuVKTps9wZUcr2BkVbaqq92HJ0" width="690" height="345" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/1/127b60e15cfa9c5dda853f318b39722cf8b05652_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">evaluate</span><span class="informations">1018×509 16.4 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div><br>
But when you use $(cat template.json) , it could only evaluate this cat command. It will not continue evaluate the json file content.<br>
<div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/8/8caf05adc2d773584a993e08e1d3a295b0d3a218.png" data-download-href="/uploads/short-url/k4xPphZ9HvAohfQi1fE1tmOFXm0.png?dl=1" title="evaluate cat" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181090585-f066ed59-2099-4ca3-83e3-dbcbc6720bcc.png" alt="evaluate cat" data-base62-sha1="k4xPphZ9HvAohfQi1fE1tmOFXm0" width="690" height="238" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/8/8caf05adc2d773584a993e08e1d3a295b0d3a218_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">evaluate cat</span><span class="informations">820×283 11.4 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div><br>
I would recommend you using sed to replace the value in json file, the context could expand in bash scripts.  And then you cat json file content. Please see my code snippet :
<pre><code class="lang-auto">      - id: gettemplate
        run: |
           sed -i '/text/c\   \"text\" : \"*${{ github.workflow }}*\n`${{ job.status }}`",' template.json
           message=$(cat template.json)
           echo ::set-output name=msg::$message
</code></pre>

**Comment by Ajibola** [comment, marked as answer]
<p>Thanks <a class="mention" href="/yanjingzhu">@yanjingzhu</a> for the detailed response, it was really helpful. Seeing that the expansion is essentially a format() call, I used that idea to substitute in the placeholders.</p>
<pre><code class="lang-auto">env:
        MSG_TEMPLATE: ${{ format(steps.gettemplate.outputs.msg, github.workflow, job.status) }}
</code></pre>
<p>The json needed to be escaped in order for format() to work as per the docs, so it ended up looking like this</p>
<blockquote>
<p>{{“text”:“Workflow Results for : {0} {1}”}}</p>
</blockquote>
<p>I think it might be better using sed to escape the curly braces in a json automatically and using format() for the substitution</p>
<p>How were you able to see the debug info?, it is quite handy for knowing what is happening under the hood.</p>

**Comment by Yanjingzhu** [comment]
<p><a class="mention" href="/ajibola">@Ajibola</a> You could also enable step debug logging for your repo by adding <code>ACTIONS_STEP_DEBUG</code> to <code>true</code> in <code>secrets</code>.<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/6/681f81ab6ddc65fa7f154b2366d25560a3f08405.png" data-download-href="/uploads/short-url/eR79SuUNPrnrpOxmBWTTwh7I3qZ.png?dl=1" title="step log" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181090586-a7140291-bed0-4037-b09c-d5e3086d9c28.png" alt="step log" data-base62-sha1="eR79SuUNPrnrpOxmBWTTwh7I3qZ" width="690" height="305" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/6/681f81ab6ddc65fa7f154b2366d25560a3f08405_2_690x305.png, https://aws1.discourse-cdn.com/github/optimized/2X/6/681f81ab6ddc65fa7f154b2366d25560a3f08405_2_1035x457.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/6/681f81ab6ddc65fa7f154b2366d25560a3f08405.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/6/681f81ab6ddc65fa7f154b2366d25560a3f08405_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">step log</span><span class="informations">1040×460 32.6 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div><br>
Please refer to this document: <a href="https://help.github.com/en/actions/configuring-and-managing-workflows/managing-a-workflow-run#enabling-step-debug-logging" rel="nofollow noopener">https://help.github.com/en/actions/configuring-and-managing-workflows/managing-a-workflow-run#enabling-step-debug-logging</a>
